### PR TITLE
Contest entity

### PR DESCRIPTION
**Pull request by sanyavertolet**
### What's done:
- [x] Added contest entity
- [x] Added ContestRepository.kt
- [x] Added ContestService.kt
- [x] Added ContestController.kt with simple endpoints
- [x] Added contest table in db
- [x] Added dev db presets for contest table
- [x] Added LnkContestProject entity
- [x] Added lnk_contest_project table in db
- [x] Added dev db presets for lnk_contest_project table
- [x] Added LnkContestProjectRepository.kt

(#250)

### CODE REVIEW

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/entities/Contest.kt [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * @property name organization
+ * @property status
+ * @property startTime the time contest starts
+ * @property endTime the time contest ends
+ * @property description
+ */
+@Entity
+@Serializable
```
Comment (petertrr) [inline save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/entities/Contest.kt:17]: Please don't mix `Entity` and `Serializable`, for Project it's just legacy and for `Organization` it was required in order not to refactor `Project` class.

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/entities/Contest.kt [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+@Suppress("USE_DATA_CLASS")
+data class Contest(
+    var name: String,
```
Comment (petertrr) [inline save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/entities/Contest.kt:20]: Should we also add category similar to benchmarks?

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ContestService.kt [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+     * @param pageSize amount of contests that should be taken
+     * @return list of active contests
+     */
+    @Suppress("MagicNumber")
+    fun findContestsInProgress(pageSize: Int? = null) = LocalDateTime.now().let {
+        contestRepository.findByStartTimeBeforeAndEndTimeAfterAndStatus(
+            it,
+            it,
+            ContestStatus.CREATED,
+            Pageable.ofSize(pageSize ?: 10),
```
Comment (petertrr) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ContestService.kt:42]: Why not put default value as default argument value?

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ContestService.kt [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        )
+    }.content ?: emptyList()
```
Comment (petertrr) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ContestService.kt:44]: Can `content` be `null`?

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+internal class ContestController(
+    private val contestService: ContestService,
+) {
+    /**
+     * @param contestName
+     * @return Organization
+     */
+    @GetMapping("/{contestName}")
+    @PreAuthorize("permitAll()")
+    fun getContestByName(@PathVariable contestName: String) = Mono.fromCallable {
```
Comment (petertrr) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt:29]: Our extension `justOrNotFound`?

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    }.switchIfEmpty {
+        Mono.error(ResponseStatusException(HttpStatus.NOT_FOUND))
+    }
+
+    /**
+     * @param pageSize amount of contests that should be taken
+     * @param authentication an [Authentication] representing an authenticated request
+     * @return list of organization by owner id
+     */
+    @GetMapping("/get/active")
```
Comment (petertrr) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt:40]:
```suggestion
    @GetMapping("/active")
```

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    ): Flux<Contest> = Flux.fromIterable(
+        contestService.findContestsInProgress(pageSize)
+    )
+
+    /**
+     * @param pageSize amount of contests that should be taken
+     * @param authentication an [Authentication] representing an authenticated request
+     * @return list of organization by owner id
+     */
+    @GetMapping("/get/finished")
```
Comment (petertrr) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt:54]:
```suggestion
    @GetMapping("/finished")
```

**Review by petertrr — CHANGES_REQUESTED** [review]

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/entities/Contest.kt (hunk 1 above)
Comment (sanyavertolet) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/entities/Contest.kt:17]: Ok, will move `Contest` entity to `jvmMain` section then and will remove `Serializable`.

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/entities/Contest.kt (hunk 2 above)
Comment (sanyavertolet) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/entities/Contest.kt:20]: I guess we will need it later, but for now it seems ok

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ContestService.kt (hunk 3 above)
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ContestService.kt:42]: Oh sure

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ContestService.kt (hunk 4 above)
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ContestService.kt:44]: Seems like no, will fix it

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt [hunk 8]
```diff
@@ -26,11 +23,7 @@ internal class ContestController(
      */
     @GetMapping("/{contestName}")
     @PreAuthorize("permitAll()")
-    fun getContestByName(@PathVariable contestName: String) = Mono.fromCallable {
-        contestService.findByName(contestName)
-    }.switchIfEmpty {
-        Mono.error(ResponseStatusException(HttpStatus.NOT_FOUND))
-    }
+    fun getContestByName(@PathVariable contestName: String) = justOrNotFound(contestService.findByName(contestName))
```
Comment (petertrr) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt:26]: These endpoints are intended to be consumed by external clients, including our frontend. So they shouldn't return Java classes in Jackson's format, instead we should create and use `ContestDto` here

Comment (sanyavertolet) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt]: Done

**Review by petertrr — APPROVED** [review]

**Pull request merged**